File: src/context/spt_js/validation.js
// -------------------------------------------------------------------------------------------------------------------
//
//   Copyright (c) 2010, Southpaw Technology Inc., All Rights Reserved
//   
//   PROPRIETARY INFORMATION.  This software is proprietary to Southpaw Technology Inc., and is not to be
//   reproduced, transmitted, or disclosed in any way without written permission.
//   
// -------------------------------------------------------------------------------------------------------------------


spt.validation = {};



spt.validation.onchange_cbk = function( evt, bvr )
{
    // assumes bvr.src_el is the input or select element that directly has a validation behavior
    var input_el = bvr.src_el;
    spt.validation.direct_input_element_check( input_el );
}


spt.validation.direct_input_element_check = function( input_el )
{
    var v_bvr_list = spt.behavior.get_bvrs_by_type( "validation", input_el );
    return spt.validation.check( input_el.value, v_bvr_list, input_el );
}


spt.validation.check = function( new_value, validation_bvr_list, display_target_el )
{
    if( ! validation_bvr_list || validation_bvr_list.length == 0 ) {
        return true;
    }

    var valid_flag = true;
    var v_msg = '';
    for( var c=0; c < validation_bvr_list.length; c++ ) {
        var v_bvr = validation_bvr_list[c];
        var v_js = v_bvr.cbjs_validation;
        if( v_js ) {
            var v_result = spt.validation._run_cbjs( v_js, new_value, display_target_el, v_bvr );

            if( v_result == null ) {
                // means an error occurred running the validation script!
                v_msg = 'ERROR running validation script ... see Web Client Output Log Console for details';
                valid_flag = false;
            } else {
                valid_flag = v_result;
                if( v_bvr.validation_warning ) {
                    v_msg = v_bvr.validation_warning;
                } else {
                    v_msg = "Entry not valid";
                }
            }
        }
    }
    if( ! valid_flag  ) {
        spt.validation.set_invalid_warning( display_target_el, v_msg );
        display_target_el.setAttribute( "title", v_msg );
    } else {
        spt.validation.clear_invalid_warning( display_target_el );
    }

    var table_top = display_target_el.getParent(".spt_table");
    // don't change color for table tds
    if (!table_top) {
        if (!valid_flag)
            display_target_el.setStyle('background','#A99');
        else
            display_target_el.setStyle('background','white');

    }
    // Special handling for Edit Panel ...
    //
    //
    var edit_top = display_target_el.getParent(".spt_edit_top");
    if( edit_top ) {

        // get insert button ...
        var btn_list = edit_top.getElements(".spt_button_top");
        var insert_btn = null;
        for( var c=0; c < btn_list.length; c++ ) {
            var btn = btn_list[c];
            var btn_label = btn.getElement('.spt_label').getProperty("spt_text_label");
            if( btn_label && ['Insert', 'Edit', 'Add', 'Save'].contains(btn_label) ) {
                insert_btn = btn;
                break;
            }
        }

        if( insert_btn ) {
            var failed_list = edit_top.getElements(".spt_input_validation_failed");
            if( failed_list.length ) {
                insert_btn.setStyle("visibility","hidden");
            } else {
                insert_btn.setStyle("visibility","visible");
            }

        }
    }

    return valid_flag;
}


spt.validation.has_invalid_entries = function( start_el, top_mt_search_str )
{
    var top_el = $(start_el).getParent( top_mt_search_str );
    if( ! top_el ) {
        return false;
    }

    var invalid_entries = top_el.getElements(".spt_input_validation_failed");
    if( invalid_entries.length ) {
        return true;
    }
    return false;
}


spt.validation._run_cbjs = function( validation_js_str, value, display_target_el, bvr )
{
    // First define the validation function wrapper around the validation js to be executed ...
    var fn_stmt = "var fn = function( value, display_target_el, bvr ) { " + validation_js_str + "; }";

    try {
        eval( fn_stmt );
    } catch(e) {
        spt.log_eval_error( e, fn_stmt, "spt.validation.run_cbjs", "error occurred in DEFINING validation function" );
        return null;
    }

    // Then run the validation function (defined above) to get the result of the validation ...
    var v_result = false;
    try {
        v_result = fn( value, display_target_el, bvr );
    } catch(e) {
        spt.log_eval_error( e, fn_stmt, "spt.validation.run_cbjs", "error occurred in RUNNING validation function" );
        return null;
    }

    return v_result;
}


spt.validation.set_invalid_warning = function( display_target_el, warning_msg )
{
    // this if statement is kept for backward compatibility only
    if( ! spt.css.has_look( "input_validation_failed", display_target_el ) ) {
        spt.css.add_looks( "input_validation_failed", display_target_el );
    }
    
    display_target_el.setStyle('background','#A99');
    display_target_el.setProperty( "title", warning_msg );
    display_target_el.addClass( "tactic_tip" );

    display_target_el.addClass( "spt_input_validation_failed" );
    display_target_el.setProperty("spt_validation_failed_msg", warning_msg);  // needed for DG-table td validations

}


spt.validation.clear_invalid_warning = function( display_target_el )
{
    if( spt.css.has_look( "input_validation_failed", display_target_el ) ) {
        spt.css.remove_looks( "input_validation_failed", display_target_el );
    }
    display_target_el.setProperty( "title", "" );
    display_target_el.removeClass( "tactic_tip" );

    display_target_el.removeClass( "spt_input_validation_failed" );

}

File: src/tactic/ui/app/validation.py
###########################################################
#
# Copyright (c) 2010, Southpaw Technology
#                     All Rights Reserved
#
# PROPRIETARY INFORMATION.  This software is proprietary to
# Southpaw Technology, and is not to be reproduced, transmitted,
# or disclosed in any way without written permission.
#
#
#
__all__ = ["ValidationUtil"]

import os
from pyasm.biz import Project
from pyasm.common import TacticException
from pyasm.search import Search



class ValidationUtil:
    SCHEME = {'INTEGER': 'return value ? spt.input.is_integer(value): true',
          'POS_INTEGER': 'return value ? spt.input.is_integer(value) && parseInt(value, 10) > 0 : true',
          'NUMERIC': 'return value ? spt.input.is_numeric(value): true' }

    def __init__(self, **kwargs):

        self.validation_regex = ''
        self.validation_script = ''
        self.validation_script_path = ''
        self.validation_js = ''
        self.validation_scheme = ''
        self.validation_warning = ''
        self.direct_cbjs = ''

        widget = kwargs.get('widget')
        if widget and hasattr(widget, "get_option"):
            self.validation_regex = widget.get_option("validation_regex")
            self.validation_script = widget.get_option("validation_script")
            self.validation_script_path = widget.get_option("validation_script_path")
            self.validation_js = widget.get_option("validation_js")
            self.validation_scheme = widget.get_option('validation_scheme')
            self.validation_warning = widget.get_option("validation_warning")

        if kwargs.get('regex'):
            self.validation_regex = kwargs.get('regex')

        if kwargs.get('script'):
            self.validation_script = kwargs.get('script')
        elif kwargs.get('script_path'):
            self.validation_script_path = kwargs.get('script_path')

        if kwargs.get('js'):
            self.validation_js = kwargs.get('js')
        elif kwargs.get('scheme'):
            self.validation_scheme = kwargs.get('scheme')

        if kwargs.get('warning'):
            self.validation_warning = kwargs.get('warning')

        if kwargs.get('direct_cbjs'):
            self.direct_cbjs = kwargs.get('direct_cbjs')

       


    def get_validation_bvr(self):

        cbjs_validation = ''

        if self.direct_cbjs:
            cbjs_validation = self.direct_cbjs

        elif self.validation_script:

            # get the script code from the custom_script table of the project ...
            prj_code = Project.get_project_code()
            s_key = "config/custom_script?project=%s&code=%s" % (prj_code, self.validation_script)
            custom_script_sobj = Search.get_by_search_key( s_key )
            if not custom_script_sobj:
                raise TacticException( 'Did NOT find a code="%s" custom_script entry for validation behavior' %
                                        (self.validation_script) )
            cbjs_validation = custom_script_sobj.get_value('script')
        
        elif self.validation_script_path:
            folder = os.path.dirname(self.validation_script_path)
            title = os.path.basename(self.validation_script_path)
            expr = "@GET(config/custom_script['folder','%s']['title','%s'].script)" %(folder, title)
            cbjs_validation = Search.eval(expr, single=True)
            

        elif self.validation_regex:
            # NOTE: the return MUST be true or false ... a return of null means an error
            # on defining or running the validation function wrapper around this ...
            #
            cbjs_validation = '''return( value.match( /%s/ ) != null );''' % self.validation_regex

        elif self.validation_js:
            cbjs_validation = self.validation_js

        elif self.validation_scheme:
            js = self.SCHEME.get(self.validation_scheme)
            cbjs_validation = js
            
        if not cbjs_validation:
            return None

        if not self.validation_warning:
            self.validation_warning = 'Entry not valid'

        return {'type': 'validation', 'cbjs_validation': cbjs_validation,
                'validation_warning': self.validation_warning}


    def get_input_onchange_bvr(self):

        return { 'type': 'change', 'cbjs_action': 'spt.validation.onchange_cbk( evt, bvr );' }


